doc/Architecture/Compilation.org

```diff
@@ -0,0 +1,62 @@
+* Compilation
+** Michelson
+Michelson is an interpreted stack based language on-top of the Tezos
+chain.
+
+Since we are compiling to Michelson, some interesting considerations
+must be made
+*** Push/Enter vs Eval/Apply
+Although the drawbacks of Push/Enter compared to Eval/Apply are well
+known [fn:1] [fn:2], for an interpretive stack based languages, the
+caveats in that statement go away.
+1. _Eval/Apply is easier to compile to a normal C like language_
+   - Michelson is a stack based language, this point does not apply
+2. _Push/Enter spread complexity around when trying to make a fast compiler_
+   - Michelson is an interpreted language, and in particular, creating
+     lambdas (closures) for over compilation is quite costly. What
+     would be cheaper for this language would be to simply push the
+     arguments onto the stack for the new function to consume, and
+     inline said function.
+*** Compilation Strategy
+Currently annotated Core is translated directly into Michelson
+
+This is less than ideal for more than a few reasons, and will be
+changed at a current release.
+
+However until that time, we will document what is *currently* being
+done and how to improve it
+
+One last note is that names are assumed to not be overlapping, as this
+happens higher up in the compilation process
+**** Virtual Stack
```

Comment:
Separately, we should talk about how we track the (real) Michelson stack while compiling (currently in order to map names to stack positions & look up types of things)

Reply:
Sure will do